### ISSUE TITLE
[integration_test] - When one test fails then execution of the next tests is stopped

### ISSUE THREAD

**Issue by HubertArciszewski95** [flutter/flutter#75216]
Hello, I was testing the new integration_test package, and I discover this issue.

**Preconditions**
1. I download the flutter demo app that is in flutter docs (the counter app)
2. I use this package [intgration_test](https://pub.dev/packages/integration_test)
3. I set up it according to the [flutter doc](https://flutter.dev/docs/testing/integration-tests)

You can find this configured demo project [HERE](https://github.com/HubertArciszewski95/flutter_integration_test_package) on my Github. 

**Steps**
1. Create two tests
2. Add an assertion that will fail on the first test
3. Run tests (_flutter drive --driver=test_driver/integration_test_driver.dart --target=test_driver/integration_test/example_test.dart_)

**Current**
When the first test fails then test execution is stopped. 
The second test doesn't run.

**Expected**
When the first test fails then an error is printed to the console and the second test runs.

Info I get in console when the first test fails
```
I/flutter (26321): (The following exception is now available via WidgetTester.takeException:)
I/flutter (26321): ══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
I/flutter (26321): The following TestFailure object was thrown running a test:
I/flutter (26321):   Expected: exactly one matching node in the widget tree
I/flutter (26321):   Actual: _TextFinder:<zero widgets with text "2" (ignoring offstage widgets)>
I/flutter (26321):    Which: means none were found but one was expected
I/flutter (26321): Assertion failed on first test
I/flutter (26321):
I/flutter (26321): When the exception was thrown, this was the stack:
I/flutter (26321): #4      main.<anonymous closure>.<anonymous closure>
(file:///C:/Users/harciszewski/Documents/VisualStudioWorkspace/flutter_integration_test_package/test_driver/integration_test/example_test.dart:24:7)     
I/flutter (26321): <asynchronous suspension>
I/flutter (26321): <asynchronous suspension>
I/flutter (26321): (elided one frame from package:stack_trace)
I/flutter (26321): ...
I/flutter (26321):
I/flutter (26321): This was caught by the test expectation on the following line:
I/flutter (26321):
file:///C:/Users/harciszewski/Documents/VisualStudioWorkspace/flutter_integration_test_package/test_driver/integration_test/example_test.dart line 24    
I/flutter (26321): The test description was:
I/flutter (26321):   tap on the floating action button; verify counter
I/flutter (26321): ════════════════════════════════════════════════════════════════════════════════════════════════════
I/flutter (26321): (If WidgetTester.takeException is called, the above exception will be ignored. If it is not, then the above exception will be dumped  
when another exception is caught by the framework or when the test ends, whichever happens first, and then the test will fail due to having not caught orexpected the exception.)
I/flutter (26321): ══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
I/flutter (26321): The following TestFailure object was thrown running a test:
I/flutter (26321):   Expected: exactly one matching node in the widget tree
I/flutter (26321):   Actual: _TextFinder:<zero widgets with text "2" (ignoring offstage widgets)>
I/flutter (26321):    Which: means none were found but one was expected
I/flutter (26321): Assertion failed on first test
I/flutter (26321):
I/flutter (26321): When the exception was thrown, this was the stack:
I/flutter (26321): #4      main.<anonymous closure>.<anonymous closure>
(file:///C:/Users/harciszewski/Documents/VisualStudioWorkspace/flutter_integration_test_package/test_driver/integration_test/example_test.dart:24:7)     
I/flutter (26321): <asynchronous suspension>
I/flutter (26321): <asynchronous suspension>
I/flutter (26321): (elided one frame from package:stack_trace)
I/flutter (26321): ...
I/flutter (26321):
I/flutter (26321): This was caught by the test expectation on the following line:
I/flutter (26321):
file:///C:/Users/harciszewski/Documents/VisualStudioWorkspace/flutter_integration_test_package/test_driver/integration_test/example_test.dart line 24    
I/flutter (26321): The test description was:
I/flutter (26321):   tap on the floating action button; verify counter
I/flutter (26321): ════════════════════════════════════════════════════════════════════════════════════════════════════
I/flutter (26321): 00:02 +0: end-to-end test tap on the floating action button; verify counter [E]
I/flutter (26321):   Test failed. See exception logs above.
I/flutter (26321):   The test description was: tap on the floating action button; verify counter
I/flutter (26321):
I/flutter (26321): 00:02 +0 -1: end-to-end test test 2
Failure Details:
Failure in method: tap on the floating action button; verify counter
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞═════════════════
The following TestFailure object was thrown running a test:
  Expected: exactly one matching node in the widget tree
  Actual: _TextFinder:<zero widgets with text "2" (ignoring
offstage widgets)>
   Which: means none were found but one was expected
Assertion failed on first test

When the exception was thrown, this was the stack:
#4      main.<anonymous closure>.<anonymous closure> (file:///C:/Users/harciszewski/Documents/VisualStudioWorkspace/flutter_integration_test_package/test_driver/integration_test/example_test.dart:24:7)
<asynchronous suspension>
<asynchronous suspension>
(elided one frame from package:stack_trace)

This was caught by the test expectation on the following line:
  file:///C:/Users/harciszewski/Documents/VisualStudioWorkspace/flutter_integration_test_package/test_driver/integration_test/example_test.dart line 24  
The test description was:
  tap on the floating action button; verify counter
═════════════════════════════════════════════════════════════════

end of failure 1
```

**Comment by OndrejKunc** [comment]
I can confirm this issue.
I was even able to reproduce it in the [example](https://github.com/flutter/plugins/tree/master/packages/integration_test/example) project of the `integration_test` package on the latest master branch just by adding a second test and making the first one fail.
 
Also, when there is some non-fatal exception in the app, like some overflow, then the test will also fail even though the test is otherwise correct. This will prevent running the second test as well.

One more issue closely related to this one is with the `FlutterError.onError` method.
When overridden, for example with the following code: `FlutterError.onError = (_) => Null;` and at the same time making the test fail, for example by `expect(true, false);` then the whole test will get stuck forever. It is already mentioned [here](https://github.com/flutter/flutter/issues/34499#issuecomment-749948889) and in #73095

Unfortunately, it looks like `integration_test` is now optimized only for the happy path scenarios and can't be used for real-world testing right now.

**Comment by darshankawar** [comment]
Issue verified and replicable on latest master.

<details>
<summary> Console error log </summary>

```
I/flutter (16149): (The following exception is now available via WidgetTester.takeException:)
I/flutter (16149): ══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
I/flutter (16149): The following TestFailure object was thrown running a test:
I/flutter (16149):   Expected: exactly one matching node in the widget tree
I/flutter (16149):   Actual: _TextFinder:<zero widgets with text "2" (ignoring offstage widgets)>
I/flutter (16149):    Which: means none were found but one was expected
I/flutter (16149): Assertion failed on first test
I/flutter (16149): 
I/flutter (16149): When the exception was thrown, this was the stack:
I/flutter (16149): #4      main.<anonymous closure>.<anonymous closure> (file:///Users/dhs/Documents/NCFlutter/flutter_integration_test_package/test_driver/integration_test/example_test.dart:24:7)
I/flutter (16149): <asynchronous suspension>
I/flutter (16149): <asynchronous suspension>
I/flutter (16149): (elided one frame from package:stack_trace)
I/flutter (16149): ...
I/flutter (16149): 
I/flutter (16149): This was caught by the test expectation on the following line:
I/flutter (16149):   file:///Users/dhs/Documents/NCFlutter/flutter_integration_test_package/test_driver/integration_test/example_test.dart line 24
I/flutter (16149): The test description was:
I/flutter (16149):   tap on the floating action button; verify counter
I/flutter (16149): ════════════════════════════════════════════════════════════════════════════════════════════════════
I/flutter (16149): (If WidgetTester.takeException is called, the above exception will be ignored. If it is not, then the above exception will be dumped when another exception is caught by the
framework or when the test ends, whichever happens first, and then the test will fail due to having not caught or expected the exception.)
I/flutter (16149): ══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
I/flutter (16149): The following TestFailure object was thrown running a test:
I/flutter (16149):   Expected: exactly one matching node in the widget tree
I/flutter (16149):   Actual: _TextFinder:<zero widgets with text "2" (ignoring offstage widgets)>
I/flutter (16149):    Which: means none were found but one was expected
I/flutter (16149): Assertion failed on first test
I/flutter (16149): 
I/flutter (16149): When the exception was thrown, this was the stack:
I/flutter (16149): #4      main.<anonymous closure>.<anonymous closure> (file:///Users/dhs/Documents/NCFlutter/flutter_integration_test_package/test_driver/integration_test/example_test.dart:24:7)
I/flutter (16149): <asynchronous suspension>
I/flutter (16149): <asynchronous suspension>
I/flutter (16149): (elided one frame from package:stack_trace)
I/flutter (16149): ...
I/flutter (16149): 
I/flutter (16149): This was caught by the test expectation on the following line:
I/flutter (16149):   file:///Users/dhs/Documents/NCFlutter/flutter_integration_test_package/test_driver/integration_test/example_test.dart line 24
I/flutter (16149): The test description was:
I/flutter (16149):   tap on the floating action button; verify counter
I/flutter (16149): ════════════════════════════════════════════════════════════════════════════════════════════════════
I/flutter (16149): 00:05 +0: end-to-end test tap on the floating action button; verify counter [E]
I/flutter (16149):   Test failed. See exception logs above.
I/flutter (16149):   The test description was: tap on the floating action button; verify counter
I/flutter (16149):   
Failure Details:
Failure in method: tap on the floating action button; verify counter
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞═════════════════
The following TestFailure object was thrown running a test:
  Expected: exactly one matching node in the widget tree
  Actual: _TextFinder:<zero widgets with text "2" (ignoring
offstage widgets)>
   Which: means none were found but one was expected
Assertion failed on first test

When the exception was thrown, this was the stack:
#4      main.<anonymous closure>.<anonymous closure> (file:///Users/dhs/Documents/NCFlutter/flutter_integration_test_package/test_driver/integration_test/example_test.dart:24:7)
<asynchronous suspension>
<asynchronous suspension>
(elided one frame from package:stack_trace)

This was caught by the test expectation on the following line:
  file:///Users/dhs/Documents/NCFlutter/flutter_integration_test_package/test_driver/integration_test/example_test.dart line 24
The test description was:
  tap on the floating action button; verify counter
═════════════════════════════════════════════════════════════════

end of failure 1



I/flutter (16149): 00:05 +0 -1: end-to-end test test 2

```

</details>

<details>
<summary>flutter doctor -v</summary>

```

[✓] Flutter (Channel master, 1.26.0-18.0.pre.121, on Mac OS X 10.15.4 19E2269
    darwin-x64, locale en-IN)
    • Flutter version 1.26.0-18.0.pre.121 at
      /Users/dhs/documents/Fluttersdk/flutter
    • Framework revision 225e408d4f (7 hours ago), 2021-02-01 22:26:03 -0500
    • Engine revision e59457b818
    • Dart version 2.12.0 (build 2.12.0-279.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.0)
    • Android SDK at /Users/dhs/Library/Android/sdk
    • Platform android-30, build-tools 30.0.0
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[!] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.3, Build version 12C33
    ! CocoaPods 1.9.3 out of date (1.10.0 is recommended).
        CocoaPods is used to retrieve the iOS and macOS platform side's plugin
        code that responds to your plugin usage on the Dart side.
        Without CocoaPods, plugins will not work on iOS or macOS.
        For more info, see https://flutter.dev/platform-plugins
      To upgrade see
      https://guides.cocoapods.org/using/getting-started.html#installation for
      instructions.

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)

[✓] VS Code (version 1.52.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.18.1

[✓] Connected device (4 available)
    • SM A260G (mobile)            • 5200763ebcfa861f • android-arm    • Android
      8.1.0 (API 27)
    • AOSP on IA Emulator (mobile) • emulator-5554    • android-x86    • Android
      9 (API 28) (emulator)
    • macOS (desktop)              • macos            • darwin-x64     • Mac OS
      X 10.15.4 19E2269 darwin-x64
    • Chrome (web)                 • chrome           • web-javascript • Google
      Chrome 88.0.4324.96

! Doctor found issues in 1 category.


```

</details>

**Comment by cyanglaz** [comment]
@dnfield is it a fundamental limitation or something we can change?

**Comment by dnfield** [comment]
This probably depends on https://github.com/flutter/flutter/issues/50816 - @jiahaog

**Comment by cantor1616** [comment]
As a work around, for my team, I use `Directory` to iterate over every integration test file and call `flutter drive` per test. This will run every test regardless of any failures, but it takes an incredible amount of time to run all the tests as the APK rebuilds and is reinstalled before every single test.

Any update on a fix for this issue would be appreciated. Thank you.

**Comment by DevinWatson** [comment]
It seems like the driver exiting on a test failure is intended? This is the complete opposite of every other test framework I've used.  

https://github.com/flutter/flutter/blob/0d79013a81d6aeb9d7b05d09bcafad6c64c09213/packages/flutter_tools/lib/src/commands/drive.dart#L293

![image](https://user-images.githubusercontent.com/1735588/119726781-7f20b880-be3f-11eb-9a53-7e7a9437cf27.png)

**Comment by jiahaog** [comment]
I don't think this is the correct behavior. Tests should continue to run even after the first one fails. I'll take a closer look.

> It seems like the driver exiting on a test failure is intended? 

That's correct. The `testResult` you're seeing is the outcome of the entire test *suite*, not a specific failure in a single test case.

**Comment by DevinWatson** [comment]
Alright I have just attempted to recreate the issue by creating two failing tests and running them back to back and it appears to be working as expected. So it seems out situation is unique and may be self inflicted. Will update if we find the root cause.


![image](https://user-images.githubusercontent.com/1735588/119985713-029def00-bf91-11eb-8bd5-7a0d18d262f8.png)


```
import 'package:flutter_test/flutter_test.dart';
import 'package:integration_test/integration_test.dart';


void main() {
  IntegrationTestWidgetsFlutterBinding.ensureInitialized();

  testWidgets("failing test example", (WidgetTester tester) async {
    expect(2 + 2, equals(5));
  });
}
```

```
import 'package:integration_test/integration_test.dart';
import 'failing_test.dart' as failing_test;
import 'failing_test2.dart' as failing_test2;

void main() {
  IntegrationTestWidgetsFlutterBinding.ensureInitialized();

  failing_test.main();
  failing_test2.main();
}
```

**Comment by DevinWatson** [comment]
Are we potentially approaching testing wrong? Our ideal state is that we build a single apk with a target file that contains all of our E2E tests. Currently that will only work if we don't have a single failure in any test. Our work around is what @cantor1616 mentioned above which forces us to build a seperate apk for every single test which pushes our E2E pipeline step to upwards of 4 hours which is less than idea.

**Comment by cantor1616** [comment]
> Tests should continue to run even after the first one fails.

@jiahaog This is not the case. If a test fails, all processes halt and no more tests are run. However, oddly enough, I've found that if you pass the argument `--keep-app-running`, and a test fails, the emulator DOES continue running the rest of the tests, BUT the terminal output stops, which of course leaves us without any insight of test results.

**Comment by jiahaog** [comment]
This is indeed a bug – the driver process exits immediately after the first failure and may terminate the app running the tests (depending on platform). This will be fixed by #83685, apologies for the trouble. 

>  I use Directory to iterate over every integration test file and call flutter drive per test.

@cantor1616 Regarding https://github.com/flutter/flutter/issues/75216#issuecomment-827622772, I don't quite understand how iterating over every integration test file will solve the problem of not continuing with the _rest of the test script_ when a failure is encountered. Can you please elaborate more?

**Comment by HubertArciszewski95** [comment]
For all people who comment this ticket and use integration_test package in their work. Be aware of this critical issue [https://github.com/flutter/flutter/issues/80503](https://github.com/flutter/flutter/issues/80503) your tests are not reliable because of it.

**Comment by cantor1616** [comment]
> Can you please elaborate more?

@jiahaog Firstly, thank you for your work on this ([your PR](https://github.com/flutter/flutter/pull/83685). To elaborate more on the work around I have in place currently, I wrote a script named `script/run_int_test.dart` which we run with `dart script/run_int_test.dart`. This file looks something like this:

```
void main() async {
  List<FileSystemEntity> fileSystemEntityList = Directory('integration_tests/')
    .listSync(recursive: true).where((fileSystemEntity) =>
      fileSystemEntity.toString().contains('File')
      && !fileSystemEntity.path.contains('util')
    ).toList();

  List<String> failedTests = [];

  for (int i = 0; i < fileSystemEntityList.length; i++) {
    List<String> flutterDriveArgs = [
      'drive',
      '--driver=integration_tests/util/flutter_driver.dart',
      '--flavor', 'dev',
      '--target=${fileSystemEntityList[i].path}',
      '--no-sound-null-safety',
    ];

    await Process.start('flutter', flutterDriveArgs).then((process) async {
      await stdout.addStream(process.stdout);
      await stdout.addStream(process.stderr);

      if (await process.exitCode == 1) failedTests.add(fileSystemEntityList[i].path);
    });
  }

  await stdout.close();
  await stderr.close();

  if (failedTests.isEmpty) {
    print('All tests passed!');
    exit(0);
  }
  
  print('${failedTests.length} tests failed:');
  failedTests.forEach((path) => print(path));
  exit(1);
}
```

So you see, the script iterates over all of our int tests and calls `flutter drive` for each one. Then we manually keep track of failures.

**Comment by jiahaog** [comment]
@cantor1616 thanks for explaining. It seems like your script is solving another problem of running *all* the integration test scripts in a single command and is less related to this issue for early termination of tests in a single script.

We've been working on this, and it may be still rough around the edges with some missing pieces (#66264), but after Flutter 2.2 you can try using `flutter test integration_test` to replace your `script/run_int_test.dart`. It wouldn't help with the APK rebuilds yet though.

**Comment by jiahaog** [comment]
@darshankawar Unfortunately no, this is still a work in progress at this point. Documentation will definitely be added once it is ready for wider adoption :)

**Comment by cantor1616** [comment]
@jiahaog Yes, not having to manually call `.main()` of every int test from a singular file would be very nice, but this "fail and quit" behavior was far more troublesome, so I'm just happy this is resolved for now.

May I ask if this fix will make it into the next release of stable channel? And if so, when can this be expected?

**Comment by jiahaog** [comment]
Yes, the fix will make it into the next stable release. Please see https://github.com/flutter/flutter/wiki/Flutter-build-release-channels for the release cadence, it will likely take a while to reach stable since 2.2 was released not long ago. It will be helpful if you can patch the PR and verify that the issue has indeed been fixed as well in case there is another problem.

**Comment by cantor1616** [comment]
Great, thanks for that documentation.

Tested it extensively yesterday and some more this morning and everything is working as expected 👍. Thanks again.

**Comment by cantor1616** [comment]
> _all_ the integration test scripts in a single command

@jiahaog by this did you mean that all `testWidgets()` calls need to be in the same file (that is the target of `flutter drive`)? Currently our "target" file looks something like this:
```
import '/int_tests/test1.dart';
import '/int_tests/test2.dart';
import '/int_tests/test3.dart';

void main() {
  test1.main();
  test2.main();
  test3.main();
}
```
and then each test file looks like this:
```
void main() {
  testWidgets('this is test1', (tester) async {
     // test logic here
  }
}
```
and I thought this was working, but now I am noticing some odd behavior with this setup...

Everything will look normal until after about 90-100 tests have run, then the rest of them seem to all try to run at the same time and we'll see the following repeatedly for all the remaining tests (30-40 tests).
```
 'package:flutter_test/src/binding.dart': Failed assertion: line 1610 pos 12: '!inTest': is not true.
I/flutter (10692):   dart:core-patch/errors_patch.dart 46:39                _AssertionError._doThrowNew
I/flutter (10692):   dart:core-patch/errors_patch.dart 36:5                 _AssertionError._throwNew
I/flutter (10692):   package:flutter_test/src/binding.dart 1610:12          LiveTestWidgetsFlutterBinding.runTest
I/flutter (10692):   package:integration_test/integration_test.dart 203:17  IntegrationTestWidgetsFlutterBinding.runTest
I/flutter (10692):   package:flutter_test/src/widget_tester.dart 148:24     testWidgets.<fn>
I/flutter (10692):   package:test_api/src/backend/declarer.dart 199:19      Declarer.test.<fn>.<fn>
I/flutter (10692):   ===== asynchronous gap ===========================
I/flutter (10692):   package:test_api/src/backend/declarer.dart 197:7       Declarer.test.<fn>
I/flutter (10692):   ===== asynchronous gap ===========================
I/flutter (10692):   package:test_api/src/backend/invoker.dart 257:7        Invoker._waitForOutstandingCallbacks.<fn>
I/flutter (10692):   
I/flutter (10692): Consider enabling the flag chain-stack-traces to receive more detailed exceptions.
I/flutter (10692): For example, 'pub run test --chain-stack-traces'.
I/flutter (10692):   'package:flutter_test/src/binding.dart': Failed assertion: line 1639 pos 12: '_pendingFrame == null': is not true.
I/flutter (10692):   dart:core-patch/errors_patch.dart 46:39           _AssertionError._doThrowNew
I/flutter (10692):   dart:core-patch/errors_patch.dart 36:5            _AssertionError._throwNew
I/flutter (10692):   package:flutter_test/src/binding.dart 1639:12     LiveTestWidgetsFlutterBinding.postTest
I/flutter (10692):   dart:async/future.dart 218:31                     new Future.sync
I/flutter (10692):   package:test_api/src/backend/invoker.dart 236:18  Invoker.runTearDowns.<fn>.<fn>
I/flutter (10692):   package:test_api/src/backend/invoker.dart 257:15  Invoker._waitForOutstandingCallbacks.<fn>
I/flutter (10692):   package:test_api/src/backend/invoker.dart 254:14  Invoker._waitForOutstandingCallbacks.<fn>
I/flutter (10692):   dart:async/zone.dart 1428:13                      _rootRun
I/flutter (10692):   dart:async/zone.dart 1328:19                      _CustomZone.run
I/flutter (10692):   dart:async/zone.dart 1863:10                      _runZoned
I/flutter (10692):   dart:async/zone.dart 1785:10                      runZoned
I/flutter (10692):   package:test_api/src/backend/invoker.dart 254:5   Invoker._waitForOutstandingCallbacks
I/flutter (10692):   package:test_api/src/backend/invoker.dart 235:9   Invoker.runTearDowns.<fn>
I/flutter (10692):   package:test_a
```

**Comment by jiahaog** [comment]
> did you mean that all testWidgets() calls need to be in the same file (that is the target of `flutter drive`)?

No, they don't, but each `flutter drive` invocation will only run the `testWidgets` in the specific file (the `--target` parameter). I'm afraid I don't know what is wrong with the issue you're having - it will be helpful if you can produce a sample project as a repro and file a new issue so we can take a look.
